src/backend/javascript-generator.ts

```diff
@@ -0,0 +1,202 @@
+import BreakStatement from '../ast/break-statement';
+import Call from '../ast/call';
+import FunctionDeclaration from '../ast/function-declaration';
+import FunctionObject from '../ast/function-object';
+import IdentifierExpression from '../ast/identifier-expression';
+import IfStatement from '../ast/if-statement';
+import ListExpression from '../ast/list-expression';
+import NumericLiteral from '../ast/numeric-literal';
+import Parameter from '../ast/parameter';
+import Program from '../ast/program';
+import ReturnStatement from '../ast/return-statement';
+import StringLiteral from '../ast/string-literal';
+import SubscriptedExpression from '../ast/subscripted-expression';
+import UnaryExpression from '../ast/unary-expression';
+import Variable from '../ast/variable';
+import VariableDeclaration from '../ast/variable-declaration';
+import WhileStatement from '../ast/while-statement';
+import Context from '../semantics/context';
+import { Statement } from '../type-definitions/plainscript-types';
+
+const indentLevel = 2;
+
+const opConversionDictionary: { [key: string]: string } = {
+  '!=': '!==',
+  '==': '===',
+  'and': '&&',
+  'not': '!',
+  'or': '||',
+};
+
+function makeOp(op: string): string {
+  return opConversionDictionary[op] || op;
+}
+
+// jsName(e) takes any PlainScript object with an id property, such as a
+// Variable, Parameter, or FunctionDeclaration, and produces a JavaScript
+// name by appending a unique identifying suffix, such as '_1' or '_503'.
+// It uses a cache so it can return the same exact string each time it is
+// called with a particular entity.
+const jsName = (() => {
+  let lastId = 0;
+  const map = new Map();
+  return (v: Variable | FunctionDeclaration | Parameter) => {
+    if (!(map.has(v))) {
+      map.set(v, ++lastId); // eslint-disable-line no-plusplus
+    }
+    return `${v.id}_${map.get(v)}`;
+  };
+})();
+
+// This is a nice helper for variable declarations and assignment statements.
+// The AST represents both of these with lists of sources and lists of targets,
+// but when writing out JavaScript it seems silly to write `[x] = [y]` when
+// `x = y` suffices.
+function bracketIfNecessary(a: string[]) {
+  return (a.length === 1) ? `${a}` : `[${a.join(',')}]`;
+}
+
+function generateLibraryFunctions() {
+  function generateLibraryStub(name: any, params: any, body: any) {
+    const entity = Context.INITIAL.declarations[name];
+    return `function ${jsName(entity)}(${params}) {${body}}`;
+  }
+  return [
+    generateLibraryStub('print', '_', 'console.log(_);'),
+    generateLibraryStub('sqrt', '_', 'return Math.sqrt(_);'),
+  ].join('');
+}
+
+function generateBlock(block: Statement[]) {
+  return block.map((s: Statement) => `${s.gen()};`).join('');
+}
+
+Argument.prototype.gen = function(): string | void {
+  return this.expression.gen();
+};
+
+AssignmentStatement.prototype.gen = function(): string {
+  const targets = this.targets.map((t: any) => t.gen());
```

Comment:
These `any`s will go away when the `AssignmentStatement` class itself is fixed up

Reply:
the any's did indeed go away ! thank you for the diligence !